src/coreclr/vm/gcenv.ee.cpp

```diff
@@ -55,6 +55,27 @@ void GCToEEInterface::RestartEE(bool bFinishedGC)
 {
     WRAPPER_NO_CONTRACT;
 
+    // The GC may change allocation contexts during the GC using GcEnumAllocContexts() so we need to
+    // update the corresponding combined limits now that the GC is complete. Doing this within
+    // GcEnumAllocContexts() is challenging to do correctly or efficiently because multiple GC threads
+    // may enumerate and modify the allocation contexts concurrently.
```

Comment:
Previously this change was doing the updates inside of GcEnumAllocContexts(). None of the CoreCLR tests were failing but as best I can tell there was a race condition doing it that way.

Reply:
Also, `GcEnumAllocContexts` is enumeration API as its name suggests, so modifying as side-effect would be a plain broken design.

Reply:
The GC does currently modify ACs inside that enumerator. I think changing that behavior in the GC and modifying the GC<->EE interface is outside the scope of this PR.

Reply:
I do not see the problem with GC side changing the state in the enumerator callback. I was commenting on the VM side enumerator updating state as side-effect of the enumerator.